File: python/src/problems1_100/same_tree.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def is_same_tree(self, p: TreeNode | None, q: TreeNode | None) -> bool:
        pass


class Default(Problem):
    @override
    def is_same_tree(self, p: TreeNode | None, q: TreeNode | None) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        if p.val != q.val:
            return False
        return self.is_same_tree(p.left, q.left) and \
            self.is_same_tree(p.right, q.right)


SOLUTIONS: list[Problem] = [Default()]

File: typescript/src/problems1_100/same-tree.ts
import TreeNode from '../concepts/tree-node';

class Problem {
  isSameTree(p: TreeNode | undefined, q: TreeNode | undefined): boolean {
    throw new Error(`Not implemented (${p}, ${q}).`);
  }
}

class Default extends Problem {
  isSameTree(p: TreeNode | undefined, q: TreeNode | undefined): boolean {
    if (!p && !q) {
      return true;
    }
    if (!p || !q) {
      return false;
    }
    if (p.val !== q.val) {
      return false;
    }
    return this.isSameTree(p.left, q.left)
        && this.isSameTree(p.right, q.right);
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
